### PR TITLE
Project-auth by Åsa Hildebrand, Linnéa Helén & Katharina Liebig

### PR DESCRIPTION
**Pull request by katliebig**
Link to my deployed project:
FE: https://secrets-project-auth.netlify.app/
BE: https://secrets-project-auth.herokuapp.com/

My reflections on how this week's project turned out: we're happy! :)

Things I'd like to have clarified or explained in more detail: -

### CODE REVIEW

**Review by newmarkdev — APPROVED** [review]
Well done ! we don't have that much to add but guess for a final product you could add styling and some API docs 😀

File: README.md [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+The first thing we did was to create all of the endpoints and the authentication needed for those. Then we created a user model for our database. We used bcrypt and crypto to encrypt the users password. Then we built a frontend with two routes to register or log in and a main page where the secret content is shown. The user has to log in to view and post to the main page. The most challenging part was writing a reducer for adding a new secret. If we had had more time we would have added styling.
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+FE: https://secrets-project-auth.netlify.app/login
```
[the remark this answers is not in the document]
Comment (KarolCislowski) [reply README.md:15]: Too retro. Remember that you will show your project to the recruiters... and guess how high are your chances that he/she will click anything when will see something unstyled whatsoever? You are 3 person team, and there is no way that you can't find time to put some love into it.

File: backend/server.js [hunk 2]
```diff
@@ -1,29 +1,130 @@
...
+  },
+  email: {
+    type: String,
+    unique: true,
+    required: true
+  },
+  accessToken: {
+    type: String,
+    default: () => crypto.randomBytes(128).toString('hex')
+  }
```
Comment (KarolCislowski) [inline backend/server.js:30]: nice model

File: backend/server.js [hunk 3]
```diff
@@ -1,29 +1,130 @@
...
+  try {
+    const newSecret = await new Secret({
+      message
+    }).save()
+
+    res.json({ success: true, newSecret })
+  } catch (error) {
+    res.status(400).json({ success: false, message: "Invalid request", error })
+  }
+})
```
Comment (KarolCislowski) [inline backend/server.js:83]: great 👍🏻

File: frontend/src/components/Form.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    fetch(API_URL(mode), config)
+      .then(res => res.json())
+      .then(data => {
+        if (data.success === true) {
+          batch(() => {
+            dispatch(user.actions.setUsername(data.username))
+            dispatch(user.actions.setAccessToken(data.accessToken))
+            dispatch(user.actions.setErrors(null))
+          })
```
Comment (KarolCislowski) [inline frontend/src/components/Form.js:40]: Nice 👍🏻

File: frontend/src/components/Main.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        if (data.success) {
+          batch(() => {
+            dispatch(secrets.actions.setSecrets(data.secrets))
+            dispatch(secrets.actions.setErrors(null))
+          })
+        } else {
+          dispatch(secrets.actions.setErrors(data))
+        }
+      })
+  }, [accessToken, dispatch])
```
Comment (KarolCislowski) [inline frontend/src/components/Main.js:44]: the second useEffect is for the situation when you have accessToken, so why do not input logic from it as the `else` block into the first one? 

```jsx
useEffect(() => {
    if (!accessToken) {
      history.push('/login')
    } else {
      const config = { ... }
      ...
    }
  }, [accessToken, history])
```

File: frontend/src/reducers/secrets.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    setErrors: (store, action) => {
+      store.errors = action.payload
+    },
+    addNewSecret: (store, action) => {
+      store.items = [action.payload, ...store.items]
+    }
+  }
+})
+
+export default secrets
```
Comment (KarolCislowski) [inline frontend/src/reducers/secrets.js:22]: really nice 👍🏻

File: frontend/src/reducers/user.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    setAccessToken: (store, action) => {
+      store.accessToken = action.payload
+    },
+    setErrors: (store, action) => {
+      store.errors = action.payload
+    }
+  }
+})
+
+export default user
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:23]: You need to store your data in the local storage to avoid logging out your user after refreshing the page.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
It is really nice app but you need to style it and persist your user in the local storage to avoid refreshing.


Because your graduation is so soon please let me know on slack (a short note with a link to this PR) when you finish fixing your app.

**Review by KarolCislowski — APPROVED** [review]
Now it is beautiful 😻

**Pull request closed without merging**